modeshape-jcr/src/main/java/org/modeshape/jcr/shell/Interpreter.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * ModeShape (http://www.modeshape.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.modeshape.jcr.shell;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import javax.jcr.Session;
+import javax.management.MBeanServer;
+import javax.management.ObjectName;
+import org.modeshape.jcr.JcrRepository;
+import org.modeshape.jcr.shell.cmd.Command;
+import org.modeshape.jcr.shell.cmd.Commands;
+
+/**
+ *
+ * @author kulikov
+ */
+public class Interpreter {
+
+    private final JcrRepository repository;
+    private LocalSession session;
+    
+    private String path;
+    
+    private final Logger log = Logger.getLogger("Shell");
+    
+    public Interpreter(JcrRepository repository) {
+        this.repository = repository;
+    }
+
+    public void start() throws IOException {
+        String repoName = repository.getName().replaceAll(" ", "_");
+        try {
+            MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
+            ObjectName name = new ObjectName("org.modeshape:type=" + repoName);
+            Shell mbean = new Shell(this);
+            mbs.registerMBean(mbean, name);
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+    }
+
+    public void stop() {
+    }
+
+    public String execute(String cmd) throws Exception {
+        log.severe("Executing " + cmd);
+        Command command = Commands.lookup(cmd);
+        command.prepare(cmd);
+        return command.perform(this);
+    }
+    
+    public String onCommand(String name, List<String> args, Map<String, String> options) throws Exception {
+        Command command = Commands.lookup(name);
+        command.prepare(args, options);
+        log.severe("Selected command " + command.name());
+        return command.perform(this);
+    }
+        
+    
+    public JcrRepository getRepository() {
+        return repository;
+    }
+    
+    public Session getJcrSession() {
+        return session.jcrSession();
+    }
+    
+    public void setJcrSession(Session session) {
```

Comment:
I'm not convinced this method should be here; i.e. I don't think it makes sense to pass a `session` from the outside. 

If I understand correctly, typical usages looks like:
```
shell.setJcrSession(shell.getRepository().login(workspace));
```
so even though this class has a reference to the `repository`, code is attempting to set a `session` by calling `login` on the repository 

IMO this should be encapsulated better and this class should have methods which get a new session instance without requiring it to be passed from the outside. For example:
```
login(String workspace)
login(String workspace, String username)
``` 
etc
which internally would set a new session instance.